Draft-Accepted/feedback_completer_auto_discovery.md

```diff
@@ -0,0 +1,269 @@
+# Auto Discovery for Feedback provider and Tab-Completer
+
+## Motivation
+
+Today, to enable a feedback provider or a tab-completer for a native command,
+a user has to import a module or run a script manually, or do that in their profile.
+There is no way to auto-discover a feedback provider or a tab-completer for a specific native command.
+
+As a tool author, I want to provide a feedback provider or a tab-completer along with my tool installation,
+but I don't want to mess with user's profile to make the feedback provider or tab-completer discoverable.
+
+As a user, I want my feedback providers and tab-completers for the specific native commands to be loaded lazily,
+instead of having to use my profile to load them at session startup.
+
+Relevant GitHub issues:
+- [Enable loading of Feedback providers via config](https://github.com/PowerShell/PowerShell/issues/25115)
+- [Add a way to lazily autoload argument completers](https://github.com/PowerShell/PowerShell/issues/17283)
+
+## Target Scenario
+
+For PowerShell commands (Function or Cmdlet), I presume the completion or feedback support,
+if there is any, will be from the same module.
+So, when the command becomes available, the feedback provider and/or tab-completer will become available too.
+
+Therefore, the auto-discovery for feedback provider and tab-completer targets native commands only.
+
+## Goals
+
+- A tool can deploy its feedback provider and/or tab-completer without needing to update a file at a central location, such as the user's profile.
+- A tool can remove its deployment cleanly without needing to update a file at a central location.
+- PowerShell can discover feedback providers and tab-completers automatically, and load one based on the right trigger.
+- A user can enable or disable the auto-discovery for a feedback provider or tab-completer.
+
+## Specification
+
+The proposal is to adopt the existing mechnism used in Bash, Zsh and Fish's completion system --
+have directories contain individual completion scripts for various commands and applications,
+whose file names should match names of the commands.
+Those completion scripts are loaded only when their corresponding commands' completion is triggered for the first time.
+
+We will have separate directories for feedback providers and tab-completers, for 2 reasons:
```

Comment:
This will make it not possible for an Appx app to participate since they can't place files outside of their own installed folder

Reply:
I mentioned Appx and MSIX packages in the "Discussion Points" below. I think for those apps:
- If they want to provide tab-completer, then the tab-completer needs to be exposed by running the tool with a special flag, such as `<tool> --ps-completion`, then the user can manually create the "deployment+configuration" for the tool using the output script. Or, even better, the user can just run the output script, which will create the deployment automatically.
- If they want to provide feedback provider, I'm not sure how that will be possible. According to https://github.com/PowerShell/PowerShell/issues/17283#issuecomment-1522133126, the tool's DLL should not be used by another process, but feedback provider and predictor are binary implementation only.

Reply:
@daxian-dbw I believe this has been resolved so that we can enable Appx packaged commands to expose PS integration, can you update?